RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronSeedProducer.cc

```diff
@@ -0,0 +1,318 @@
+    
+    // Create PreId
+    reco::PreId preid;
+
+    // Add track ref to PreId
+    preid.setTrack(track_ref);
+
+    // Add variables related to track-ECAL matching to PreId
+    propagate_to_ecal(preid,track_ref,clusters);
+
+    // Add variables related to GSF tracks to PreId
+    light_gsf_tracking(preid,track_ref,seed,setup); 
+
+    // Decision based on BDT 
+    bool bdt = pass_through_ || bdt_decision(preid);
+
+    // Store PreId
+    preids->push_back(preid);
+
+    // 
+    if ( !bdt ) { continue; }
+    
+    seeds->push_back(seed);
+
+  }
+
+  event.put(std::move(seeds));
+  event.put(std::move(preids));
+
+}
+
+bool LowPtGsfElectronSeedProducer::propagate_to_ecal( reco::PreId& preid,
+						      const reco::TrackRef& track_ref,
+						      const edm::Handle<PFClusters>& clusters )
+{
+
+  // Propagate 'electron' to ECAL surface
+  float energy = sqrt( pow(0.000511,2.) + track_ref->outerMomentum().Mag2() );
+  XYZTLorentzVector mom = XYZTLorentzVector( track_ref->outerMomentum().x(),
+					     track_ref->outerMomentum().y(),
+					     track_ref->outerMomentum().z(),
+					     energy );
+  XYZTLorentzVector pos = XYZTLorentzVector( track_ref->outerPosition().x(),
+					     track_ref->outerPosition().y(),
+					     track_ref->outerPosition().z(),
+					     0. );
+  math::XYZVector field(field_->inTesla(GlobalPoint(0,0,0)));
+  BaseParticlePropagator particle( RawParticle(mom,pos), 0, 0, field.z() );
+  particle.setCharge(track_ref->charge());
+  particle.propagateToEcalEntrance(false);
+  if ( particle.getSuccess() != 0 ) { return false; }
+  
+  // ECAL entry point for track
+  GlobalPoint ecal_pos(particle.vertex().x(),
+		       particle.vertex().y(),
+		       particle.vertex().z());
+  // Preshower limit
+  bool below_ps = pow(ecal_pos.z(),2.) > pow(2.50746495928f,2.)*ecal_pos.perp2();
+
+  // Store info for PreId
+  struct Info {
+    reco::PFClusterRef clu_ref = reco::PFClusterRef();
+    float min_dr = 1.e6;
+    float dr = 1.e6;
+    float deta = 1.e6;
+    float dphi = 1.e6;
+    math::XYZPoint shower_pos = math::XYZPoint(0.,0.,0.);
+  } info;
+  
+  // Iterate through ECAL clusters 
+  for ( unsigned int iclu = 0; iclu < clusters.product()->size(); iclu++ ) {
+    reco::PFClusterRef clu_ref(clusters,iclu);
+
+    // Correct ecal_pos for shower depth 
+    double shower_depth = reco::PFCluster::getDepthCorrection(clu_ref->correctedEnergy(),
+							      below_ps,
+							      false);
+    GlobalPoint shower_pos = ecal_pos + 
+      GlobalVector(particle.momentum().x(),
+		   particle.momentum().y(),
+		   particle.momentum().z()).unit() * shower_depth;
+
+    // Determine deta, dphi, dr
+    float deta = std::abs( clu_ref->positionREP().eta() - shower_pos.eta() );
+    float dphi = std::abs( clu_ref->positionREP().phi() - shower_pos.phi() );
+    if ( dphi > float(TMath::Pi())) { dphi -= float(TMath::TwoPi()); }
+    float dr = std::sqrt( std::pow(dphi,2.f) + std::pow(deta,2.f) );
```

Comment:
please use an existing dR routine: https://github.com/cms-sw/cmssw/blob/master/DataFormats/Math/interface/deltaR.h

also its better to calculate dR2 rather than dR, it saves a sqrt. 

finally as a comment, std::pow() to square anything is inefficient. Sure the compiler may optimise it away but dphi*dphi ensures you get the fast version.

Reply:
Done, [here](https://github.com/CMSBParking/cmssw/pull/3/commits/a3b1651122bc8e5d42aa99766b81b657568356d9#diff-8ba575558221aacbec6290253b60358cR166)